### PR TITLE
[GOBBLIN-1783] Initialize scheduler with batch gets instead of individual get per flow

### PR DESCRIPTION
**Pull request by umustafi**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1783 

### Description
- [X] Here are some details about my PR, including screenshots (if applicable):
We seek to improve initialization time of the `GobblinServiceJobScheduler` upon restart or new leadership change by batching the mysql queries to get flow specs. Instead of making 1 mysql get call for each flow execution id, which scales extremely poorly with number of flows, we should group them to reduce number of calls and downtime.

This implementation adds two new functions to the `SpecStore` interface, `getSortedSpecURIs` and `getBatchedSpecs`, that we use to achieve the batching. Because these two functionalities are generic enough to be used in derived classes of the `SpecStore` we add them to the base class. Although this requires any child classes to implement these functions, it allows any consumer of the parent class `SpecStore` to use this functionality without caring about the specific implementation of the `SpecStore` used (as `GobblinServiceJobScheduler` does). Additionally, the `getBatchedSpecs` requires an `offset` or starting point to obtain the batches from so the consumer has to do some book keeping of where in the paginated gets we are but this again separates the functionality from the use case of the consumer. the entirety of the flow catalog is too large to load into memory for the `GobblinServiceJobScheduler`, so we use this batch functionality. 

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
The main testing for this functionality will be empirical testing to determine the batch size which results in decreasing the time needed for all the gets above. In general, the time should go down with a larger batch size but there may be an inflection point where we see diminishing returns or the memory required to store a large batch is excessive. I propose testing with batch size 100, 250, and 750 incrementally and using the time metrics emitted to settle on an appropriate size.

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/MysqlBaseSpecStore.java [hunk 1]
```diff
@@ -84,6 +84,7 @@ public class MysqlBaseSpecStore extends InstrumentedSpecStore {
   private static final String GET_ALL_STATEMENT = "SELECT spec_uri, spec FROM %s";
   private static final String GET_ALL_URIS_STATEMENT = "SELECT spec_uri FROM %s";
   private static final String GET_ALL_URIS_WITH_TAG_STATEMENT = "SELECT spec_uri FROM %s WHERE tag = ?";
+  private static final String GET_SPECS_BATCH_STATEMENT = "SELECT spec_uri, spec FROM %s ORDER BY spec_uri ASC LIMIT ? OFFSET ?";
```
Comment (AndyJiang99) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/MysqlBaseSpecStore.java:87]: It's possible that we can have two specs with the same spec_uri I believe. We should keep the modified_time ordering part of the statement too so if there are two specs with the same spec_uri so the ordering of the result from the same query stays the same at all times

**Review by ZihanLi58 — COMMENTED** [review]
quickly go through the change, and two questions:
1. what will happen if we add one spec during the process you load spec into memory? Will you ends up with missing one spec finally? 
2. what happens if spec consumer try to call onAdd(flowSpecA) and then initialize method call onAdd(flowSpecA) again? Should we use the modification time to avoid this situation?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/MysqlBaseSpecStore.java (hunk 1 above)
Comment (umustafi) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/MysqlBaseSpecStore.java:87]: the `spec_uri` is the primary key [here](https://jarvis.corp.linkedin.com/codesearch/result/?name=MysqlBaseSpecStore.java&path=gobblin-elr%2Fgobblin-runtime%2Fsrc%2Fmain%2Fjava%2Forg%2Fapache%2Fgobblin%2Fruntime%2Fspec_store&reponame=linkedin%2Fgobblin-elr#91) in `MysqlBaseSpecStore` and in the `MysqlSpecStore` so ordering on spec_uri should be unique. Let me add a comment in the docstring of the paginate function.

**Comment by umustafi** [comment]
> 

1. Current implementation, adds scheduler then the specConsumer to list of services. I considered switching order but scheduler needs to be initialized before consuming specs and trying to add to scheduler. Need to confirm if services are initialized in that other or done concurrently. specConsumer starts consuming from the latestOffset so this should not miss any specs. The offset won't move along unless service is up and able to accept requests and our consumer is processing. 

2. The problem can come up if we are loading flowSpecA from old value and while processing that batch there's API request to update flow and consumer calls onAdd with a newer value first, then scheduler calls with old value. It's very rare but we may want to add modified timestamp to avoid. This technically _could_ have happened in previous case although much more rare chance with the individual gets that in between get and add spec the consumer processed a newer spec version. If we want to use modification time, need to make bigger change to store modified time with spec in `DagManager` or `Scheduler` itself perhaps.

**Comment by umustafi** [comment]
service test looks flakey with a timeout, they pass locally. Fixed checkstyle (core tests), working on the module ones

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/MysqlBaseSpecStore.java (hunk 1 above)
Comment (AndyJiang99) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/MysqlBaseSpecStore.java:87]: 1. Was there any reason why spec_json was removed from the query string? 
2. Using this query, it will error in the scenario the OFFSET or LIMIT was set to a negative value as MySQL queries cannot handle those and this [if block](https://github.com/apache/gobblin/pull/3640/files#diff-900cc8e4e863a8057e0e808230a2f9c3c169048059d62a5d94ed9beb94360c12L283-L293) handles such cases. We'll need to add something similar in to handle those cases.

Comment (umustafi) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/MysqlBaseSpecStore.java:87]: Ah that's something we should specify in the function parameter values expected and throw an error if we receive input outside of our expected bounds. We shouldn't receive negative values (count = 0 or offset = 0 should be fine)

File: gobblin-runtime/src/test/java/org/apache/gobblin/runtime/spec_store/MysqlSpecStoreTest.java [hunk 2]
```diff
@@ -303,19 +303,18 @@ public void testGetAllSpecPaginate() throws Exception {
...
     Assert.assertEquals(specs.size(), 2);
     Assert.assertTrue(specs.contains(this.flowSpec1));
     Assert.assertTrue(specs.contains(this.flowSpec2));
     Assert.assertFalse(specs.contains(this.flowSpec4));
 
-    // Return all flowSpecs from index 0 to 2 - 1. Total of 3 flowSpecs, only return first two.
-    // Check that functionality for not including a start value is the same as including start value of 0
-    specs = this.specStore.getSpecs(-1, 2);
+    // Return all flowSpecs from index 1 to 3 - 1. Total of 2 flowSpecs, only return second two.
+    specs = this.specStore.getSpecsPaginated(1, 2);
```
Comment (AndyJiang99) [inline gobblin-runtime/src/test/java/org/apache/gobblin/runtime/spec_store/MysqlSpecStoreTest.java:313]: Same comment as below, can we keep the -1 input for either OFFSET or COUNT just to check that the functionality to handle such edge cases exists

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 3]
```diff
@@ -133,10 +147,22 @@ public GobblinServiceJobScheduler(@Named(InjectionNames.SERVICE_NAME) String ser
...
+    this.loadSpecsBatchSize = Integer.parseInt(ConfigUtils.configToProperties(config).getProperty(ConfigurationKeys.LOAD_SPEC_BATCH_SIZE, String.valueOf(ConfigurationKeys.DEFAULT_LOAD_SPEC_BATCH_SIZE)));
     this.isNominatedDRHandler = config.hasPath(GOBBLIN_SERVICE_SCHEDULER_DR_NOMINATED)
         && config.hasPath(GOBBLIN_SERVICE_SCHEDULER_DR_NOMINATED);
     this.warmStandbyEnabled = warmStandbyEnabled;
     this.quotaManager = quotaManager;
+    // Check that these metrics do not exist before adding, mainly for testing purpose which creates multiple instances
+    // of the scheduler. If one metric exists, then the others should as well.
+    MetricFilter filter = MetricFilter.contains(RuntimeMetrics.GOBBLIN_JOB_SCHEDULER_AVERAGE_GET_SPEC_SPEED_WHILE_LOADING_ALL_SPECS_MILLIS);
+    if (metricContext.getGauges(filter).isEmpty()) {
+//      this.averageGetSpecTimeMillis =
```
Comment (AndyJiang99) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:159]: Remove unnecessary comments here

**Review by AndyJiang99 — CHANGES_REQUESTED** [review]

File: gobblin-runtime/src/test/java/org/apache/gobblin/runtime/spec_store/MysqlSpecStoreWithUpdateTest.java [hunk 4]
```diff
@@ -325,19 +325,18 @@ public void testGetAllSpecPaginate() throws Exception {
...
     Assert.assertEquals(specs.size(), 2);
     Assert.assertTrue(specs.contains(this.flowSpec1));
     Assert.assertTrue(specs.contains(this.flowSpec2));
     Assert.assertFalse(specs.contains(this.flowSpec4));
 
-    // Return all flowSpecs of index [0, 2). Total of 3 flowSpecs, only return first two.
-    // Check that functionality for not including a start value is the same as including start value of 0
-    specs = this.specStore.getSpecs(-1, 2);
+    // Return all flowSpecs from index 1 to 3 - 1. Total of 2 flowSpecs, only return second two.
+    specs = this.specStore.getSpecsPaginated(1, 2);
```
[the remark this answers is not in the document]
Comment (umustafi) [reply gobblin-runtime/src/test/java/org/apache/gobblin/runtime/spec_store/MysqlSpecStoreWithUpdateTest.java:335]: Added tests for all the edge cases of negative, 0 count/offset or a start offset past the length of the store.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/MysqlBaseSpecStore.java [hunk 5]
```diff
@@ -72,6 +73,7 @@ public class MysqlBaseSpecStore extends InstrumentedSpecStore {
 
   public static final String CONFIG_PREFIX = "mysqlBaseSpecStore";
   public static final String DEFAULT_TAG_VALUE = "";
+  public static final String modificationTimeKey = "modified_time";
```
Comment (Will-Lo) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/MysqlBaseSpecStore.java:76]: Can we make this a Configuration Key or a key associated with the FlowSpec rather than the SpecStore? Since the property is part of the flowSpec it will encapsulate better

**Review by Will-Lo — CHANGES_REQUESTED** [review]

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 6]
```diff
@@ -210,26 +239,67 @@ private void scheduleSpecsFromCatalog() {
...
+        }
+      }
+      startOffset += this.loadSpecsBatchSize;
+      // This count is used to ensure the average spec get time is calculated accurately for the last batch which may be
+      // smaller than the loadSpecsBatchSize
+      averageGetSpecTimeValue = (batchGetEndTime - batchGetStartTime) / batchOfSpecs.size();
+    }
+
+    // Ensure no specs were more specs were added during the load time
+    int updatedNumSpecs = flowCatalog.get().getSize();
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:279]: As we talked before, if delete spec happens, this won't solve the issue, am I missing anything? Why not just use set to track what has been handled?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 7]
```diff
@@ -342,8 +412,24 @@ public AddSpecResponse onAddSpec(Spec addedSpec) {
       }
     }
 
+    // Compare the modification timestamp of the spec being added if the scheduler is being initialized, ideally we
+    // don't even want to do the same update twice as it will kill the existing flow and reschedule it unnecessarily
+    Long modificationTime = Long.valueOf(flowSpec.getConfigAsProperties().getProperty(MysqlBaseSpecStore.modificationTimeKey, "0"));
+    if (this.isSchedulingSpecsFromCatalog()) {
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:418]: we can do this compare all the time even if it's not the loading time period? What's the concern for not doing that?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 8]
```diff
@@ -342,8 +412,24 @@ public AddSpecResponse onAddSpec(Spec addedSpec) {
...
+      String uriString = flowSpec.getUri().toString();
+      // If spec does not exist in scheduler or have a modification time associated with it, assume it's the most recent
+      if (this.scheduledFlowSpecs.containsKey(uriString) && this.lastUpdatedTimeForFlowSpec.containsKey(uriString)) {
+        if (this.lastUpdatedTimeForFlowSpec.get(uriString) >= modificationTime) {
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:422]: if modification time is 0, which means it's not enabled with multi leader, I believe you want to process it anyway

**Review by ZihanLi58 — CHANGES_REQUESTED** [review]
Seems there is no change for spec monitor, I think for spec monitor, you also need to get modification time in the spec?

**Comment by umustafi** [comment]
> Seems there is no change for spec monitor, I think for spec monitor, you also need to get modification time in the spec?

Good catch, I changed the getSpec function for a single URI in SpecStore to also use modification time

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java (hunk 7 above)
Comment (umustafi) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:418]: If not scheduling period, then as of now we should always be processing most recent update but we can do the check anyway since its small.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java (hunk 8 above)
Comment (umustafi) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:422]: Added another condition for this to process if modification time is always 0 then process.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 9]
```diff
@@ -210,26 +245,67 @@ private void scheduleSpecsFromCatalog() {
...
+      // smaller than the loadSpecsBatchSize
+      averageGetSpecTimeValue = (batchGetEndTime - batchGetStartTime) / batchOfSpecs.size();
     }
+
+    // Ensure we did not miss any specs due to ordering changing (deletions/insertions) while loading
+    Iterator<URI> urisLeft = urisLeftToSchedule.iterator();
+    while (urisLeft.hasNext()) {
+        URI uri = urisLeft.next();
+        try {
+          Spec spec = this.flowCatalog.get().getSpecs(uri);
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:290]: Use getSpecWrapper? It's possible that it's adhoc flow and we delete it after execute.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 10]
```diff
@@ -210,26 +245,67 @@ private void scheduleSpecsFromCatalog() {
...
+          // Disable FLOW_RUN_IMMEDIATELY on service startup or leadership change if the property is set to true
+          if (spec instanceof FlowSpec && PropertiesUtils
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:292]: Duplicate code, can we extract them as one method?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 11]
```diff
@@ -342,8 +418,24 @@ public AddSpecResponse onAddSpec(Spec addedSpec) {
...
 
+    // Compare the modification timestamp of the spec being added if the scheduler is being initialized, ideally we
+    // don't even want to do the same update twice as it will kill the existing flow and reschedule it unnecessarily
+    Long modificationTime = Long.valueOf(flowSpec.getConfigAsProperties().getProperty(FlowSpec.modificationTimeKey, "0"));
+    String uriString = flowSpec.getUri().toString();
+    // If the modification time is 0 (which means the original API was used to retrieve spec or warm standby mode is not
+    // enabled), spec not in scheduler, or have a modification time associated with it assume it's the most recent
+    if (modificationTime != 0L && this.scheduledFlowSpecs.containsKey(uriString)
+        && this.lastUpdatedTimeForFlowSpec.containsKey(uriString)) {
+      if (this.lastUpdatedTimeForFlowSpec.get(uriString) >= modificationTime) {
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:429]: might be a super corner case, but if the modification time is the same, you might want to check whether the flow is run-immediately, as the load method overwrite that value and might skip the first run of the flow.

**Review by ZihanLi58 — CHANGES_REQUESTED** [review]
Minor comments, also we seems to change all the get method return modification time in the flow spec as well. Hopefully you verified that this is a backward compatible change

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 12]
```diff
@@ -342,8 +418,24 @@ public AddSpecResponse onAddSpec(Spec addedSpec) {
...
+        _log.info("Ignoring the spec {} modified at time {} because we have a more updated version from time {}",
+            addedSpec, modificationTime,this.lastUpdatedTimeForFlowSpec.get(uriString));
+        return new AddSpecResponse(response);
+      }
+    }
+
     // todo : we should probably not schedule a flow if it is a runOnce flow
     this.scheduledFlowSpecs.put(flowSpecUri.toString(), addedSpec);
+    this.lastUpdatedTimeForFlowSpec.put(flowSpecUri.toString(), modificationTime);
```
Comment (Will-Lo) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:438]: Do we still need to keep track of this data structure after service startup?

**Review by Will-Lo — CHANGES_REQUESTED** [review]

**Comment by umustafi** [comment]
> Minor comments, also we seems to change all the get method return modification time in the flow spec as well. Hopefully you verified that this is a backward compatible change

Tests pass (other than when you compare user generated spec to retrieved one which requires some change). Logically this should be backwards compatible, since we added a property not removed. Other Specs should also be able to add properties as needed to the Config.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java (hunk 12 above)
Comment (umustafi) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:438]: We don't have to but there's no harm in having the check regardless.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/metrics/RuntimeMetrics.java [hunk 13]
```diff
@@ -54,6 +54,9 @@ public class RuntimeMetrics {
   public static final String GOBBLIN_MYSQL_QUOTA_MANAGER_QUOTA_REQUESTS_EXCEEDED = ServiceMetricNames.GOBBLIN_SERVICE_PREFIX + "gobblin.mysql.quota.manager.quotaRequests.exceeded";
   public static final String GOBBLIN_MYSQL_QUOTA_MANAGER_TIME_TO_CHECK_QUOTA = ServiceMetricNames.GOBBLIN_SERVICE_PREFIX + "gobblin.mysql.quota.manager.time.to.check.quota";
 
+  public static final String GOBBLIN_JOB_SCHEDULER_AVERAGE_GET_SPEC_SPEED_WHILE_LOADING_ALL_SPECS_MILLIS = ServiceMetricNames.GOBBLIN_SERVICE_PREFIX + ".jobScheduler.average.get.spec.speed.while.loading.all.specs.millis";
```
[the remark this answers is not in the document]
Comment (umustafi) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/metrics/RuntimeMetrics.java:57]: good call about the naming, I update all have a prefix `jobScheduler` and removed the delimiters between the name. The prefix does not contain a `.` after the `GobblinService` so we need one before the suffix otherwise they get concatenated.

**Review by Will-Lo — APPROVED** [review]

**Review by AndyJiang99 — APPROVED** [review]

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 14]
```diff
@@ -343,8 +416,27 @@ public AddSpecResponse onAddSpec(Spec addedSpec) {
...
+    Long modificationTime = Long.valueOf(flowSpec.getConfigAsProperties().getProperty(FlowSpec.MODIFICATION_TIME_KEY, "0"));
+    String uriString = flowSpec.getUri().toString();
+    Boolean isRunImmediately = PropertiesUtils.getPropAsBoolean(jobConfig, ConfigurationKeys.FLOW_RUN_IMMEDIATELY, "false");
+    // If the modification time is 0 (which means the original API was used to retrieve spec or warm standby mode is not
+    // enabled), spec not in scheduler, or have a modification time associated with it assume it's the most recent
+    if (modificationTime != 0L && this.scheduledFlowSpecs.containsKey(uriString)
+        && this.lastUpdatedTimeForFlowSpec.containsKey(uriString)) {
+      // For run-immediately flows with a schedule the modified_time would remain the same
+      if (this.lastUpdatedTimeForFlowSpec.get(uriString) > modificationTime
+          || (this.lastUpdatedTimeForFlowSpec.get(uriString).equals(modificationTime) && !isRunImmediately)) {
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:430]: why equals but not ==?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java (hunk 12 above)
Comment (ZihanLi58) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:439]: I think you want to delete the entry when we remove the flow from the scheduler, otherwise, you will have a slight memory leak when we keep receiving ad-hoc flow requests.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java (hunk 14 above)
Comment (umustafi) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java:430]: I'm comparing `Long` object not primitive so it gives a warning about == comparison that may check the object references are equal not value

**Review by ZihanLi58 — APPROVED** [review]
+1
